feat: add user account settings page

aaronc opened the pull request:
## Description

Closes: #2156

This PR adds the `UserAccountSettings` component and styles it with a combination of existing elements and Tailwind CSS.

The Tailwind CSS configuration was done using a shared `tailwind.common.js` configuration file and `tailwind.css` in the root of the repo which each project should setup with its tailwind configuration. Currently Tailwind is only configured for Storybook so in follow-ups we will need to enable it for `web-marketplace`, `web-www`, etc.

The default Tailwind colors and spacing scales have been overriden to match our brand colors and @erikalogie's spacing in Figma desings. I have only added a small subset of colors based primarily on the naming in the dark/light Figma color pallete: https://www.figma.com/file/Md1Cl2B9eJi4NMzQgpbdMF/Dark-Mode?type=design&node-id=0-1&mode=design&t=YKxx8fYvslM2TeLp-0. In follow-ups we will need to add the remaining colors and setup dark mode. I have coded the colors using CSS vars to make dark mode easy to setup using a media query (`@media (prefers-color-scheme: dark)`) on the vars.

Storybook has also been updated to the latest version in this PR and all storybook dependencies are now in web-storybook/package.json.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] provided a link to the relevant issue or specification
- [ ] provided instructions on how to test
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### How to test

<!-- This should include steps on how to test the fixes or features within the marketplace app and/or the website.
If applicable, this should also include links to the created/updated components on storybook. -->

1.

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

aaronc commented on package.json:
```diff
@@ -66,7 +60,6 @@
     "patch-package": "^8.0.0",
     "prettier": "2.4.1",
     "shx": "^0.3.2",
-    "storybook": "^7.3.2",
```
All storybook dependencies moved to web-storybook

aaronc commented on tailwind.common.js:
```diff
@@ -0,0 +1,41 @@
+    // This has the side effect that some tailwind styles don't work as expected
+    // by default. For instance, in order for borders to show up at all,
+    // border-solid is needed whereas it would normally be set by default.
+    preflight: false,
```
Note the conflict of MUI and preflight. I think this is an okay trade off. The only place there was any place weird with tailwind was with borders as noted above

flagrede replied:
Ok, I think this is fine for now.

flagrede commented:
@aaronc thanks for setting that up. I've added a commit to enable tailwind in all packages [b5fc138](https://github.com/regen-network/regen-web/pull/2167/commits/b5fc138ecf8d621dbc74fb53a4f54da6af7406c7) and another one to add a small util to handle class overring [bbbe722](https://github.com/regen-network/regen-web/pull/2167/commits/bbbe7220d127d4d610e5b4cd87722fa889f6b9ee)

flagrede reviewed (approved):
Nice work, looks good!
As a follow-up, we need to:
- setup all colors
- install a tailwind class sorter (either prettier or eslint)

blushi commented:
@erikalogie you can check the UI for this on https://deploy-preview-2167--regen-storybook.netlify.app/?path=/story/registry-organisms-useraccountsettings--default

blushi commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,118 @@
+        </Body>
+        <Body size="sm">{connected ? 'Connected' : 'Not connected'}</Body>
+      </div>
+      <ContainedButton onClick={connect}>Connect</ContainedButton>
```
also in case we have a connected wallet provider, we should show the wallet address

blushi commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,118 @@
+          <Body size="sm" color="info.dark-grey">
+            Connect a wallet address to be able to perform on chain tasks such
+            as buying and selling credits, and creating projects on Regen
+            Ledger.
```
```suggestion
            Ledger.{' '}
```

aaronc replied:
Will `&nbsp;` work the same way @blushi ?

blushi replied:
yes it should work too, but `{' '}` was already used above and I think this is what my linter also defaults to but no preference

erikalogie commented:
@blushi I just realized I never provided a mobile version of this screen, sorry about that. https://www.figma.com/file/Sb9BavGvlCFPQFpjdDlwsL/Web-2.0-log-in-and-actions?type=design&node-id=1503%3A36920&mode=design&t=J3EbIjM7BIM1Cuh4-1

What you did looks pretty close but it would be good to double-check, especially the external margins seem really big but maybe that doesn't matter in the page context.

Let's add a "»" behind the learn more links. Also there should be a space between "Regen Ledger." and "Learn more" in the second paragraph.

I can't get the controls to work, I'm just changing false to true, etc. and nothing changes.

aaronc commented:
@erikalogie in order to properly implement the copy address button, I need different states for the copy icon - definitely a pressed state and ideally a hover state too. Otherwise, it doesn't feel like an actual button.

erikalogie commented:
> @erikalogie in order to properly implement the copy address button, I need different states for the copy icon. There might be a few ways to do this:
> 
> * implement pressed and hover states for the button
> * switch the button to say "Copied" after the address has been copied
> 
> Either way, we need some more visual interaction indicators here. Otherwise it doesn't feel like a true button and the user will feel like clicking it isn't actually working.

Oh yeah I detailed this somewhere else and just copied into this file. See here: https://www.figma.com/file/Sb9BavGvlCFPQFpjdDlwsL/Web-2.0-log-in-and-actions?type=design&node-id=1564%3A92835&mode=design&t=Sg9VVXl3bw9enUeU-1

I think the color for pressed and hover can be the same FYI.

aaronc commented:
> I think the color for pressed and hover can be the same FYI.

That works but it's not super satisfying to have no completion action. Anyway, small detail. Fine to postpone it. We could also just switch to showing the text "Copied!"

aaronc commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,116 @@
+         className="cursor-pointer hover:stroke-grey-300"
+    >
+      <path fill="#000" fill-rule="evenodd" d="M19 8v9H9.25a.25.25 0 0 1-.25-.25V3h5v4.25c0 .414.336.75.75.75H19Zm-4-4.586L18.287 7H15V3.414ZM3.25 5H7V1.25A.25.25 0 0 1 7.25 1h8.14c.07 0 .137.03.184.081l5.36 5.847a.25.25 0 0 1 .066.17V18.75a.25.25 0 0 1-.25.25H17v3.75a.25.25 0 0 1-.25.25H3.25a.25.25 0 0 1-.25-.25V5.25A.25.25 0 0 1 3.25 5ZM7 16.75V7H5v14h10v-2H7.25a.25.25 0 0 1-.25-.25v-2Z" clip-r ...
+    </svg>
```
Note the inline svg here and for the x icon above. This is necessary for tailwind styling. Should I pull these out into standalone components in web-components maybe?

aaronc replied:
Also for this copy button, I would prefer to use a headless button component rather than an on click handler for an svg element. Doesn't need to be in this PR necessarily, but I think it would be a good practice. React Aria Components has one: https://react-spectrum.adobe.com/react-aria/Button.html. @flagrede I know you expressed concern about react aria components being in alpha, now it's in beta and to me feels like a pretty solid library overall. If Radix Primitives had a button component, I would happily use that instead, but they don't (https://github.com/radix-ui/primitives/issues/892 - note the involvement of the React Aria maintainer here).

blushi replied:
> Note the inline svg here and for the x icon above. This is necessary for tailwind styling. Should I pull these out into standalone components in web-components maybe?

This is usually what we do for icons, see web-components/src/components/icons, although they all rely on MUI `SvgIcon` currently
I believe we already had some x icon btw, see "CloseIcon" in https://regen-storybook.netlify.app/?path=/story/icons--all-icons but I'm not sure if that's exactly the same here cc/ @erikalogie 
It also looks a bit darker than on figma

erikalogie replied:
The close x button is the same everywhere (but different colors though obviously)

flagrede replied:
If it's in beta now it should be safer to use. We're not taking too much with a button component anyway, my concern was more for more advanced components. I like Radix UI overall but react-aria might be a better fit for us, let's start with that.

aaronc commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,118 @@
+   * It should not be called if connected is true.
+   */
+  connect: () => void;
+}
```
Done

erikalogie commented:
> > I think the color for pressed and hover can be the same FYI.
> 
> That works but it's not super satisfying to have no completion action. Anyway, small detail. Fine to postpone it. We could also just switch to showing the text "Copied!"

Did you see the completion action in the 3rd screen? There is a green banner at the top.

aaronc commented:
Okay, I think I have addressed all the comments here including wrapping for mobile view and the toast banner for the help text (using Radix UI toast).

aaronc commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.types.ts:
```diff
@@ -0,0 +1,55 @@
+  disconnect: () => void;
+}
+
+type ConnectionState = DisconnectedState | ConnectedState;
+
+/**
+ * SocialProviderInfo describes a social provider and its connection info.
+ */
+export type SocialProviderInfo = {
+  /** The name of the social provider. */
+  providerName: string;
+} & ConnectionState
+
+type WalletConnectedState = {
+  /** The on-chain address of the wallet. Address will be undefined if the
+   * there is no wallet connected.
+   */
+  address: string;
+} & ConnectedState
+
+export type WalletProviderInfo = WalletConnectedState | DisconnectedState;
```
Note that I have structured types here using union and intersection types. It's a little more succinct and specific this way, but more advanced typescript. Let me know if anyone has an objection.

blushi commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,147 @@
+  );
+};
+
+const ConnectField = ({
```
usually we extract these subcomponents into separate files and name it `MainComponent.SubComponent.tsx`, here it would be `UserAccountSettings.ConnectField`

blushi commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,147 @@
+import { Subtitle, Body } from 'web-components/lib/components/typography';
```
could you enable your IDE linter, it looks like there are some linting issues in this file

blushi commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.tsx:
```diff
@@ -0,0 +1,147 @@
+        <path fill="#000" fillRule="evenodd" d="M19 8v9H9.25a.25.25 0 0 1-.25-.25V3h5v4.25c0 .414.336.75.75.75H19Zm-4-4.586L18.287 7H15V3.414ZM3.25 5H7V1.25A.25.25 0 0 1 7.25 1h8.14c.07 0 .137.03.184.081l5.36 5.847a.25.25 0 0 1 .066.17V18.75a.25.25 0 0 1-.25.25H17v3.75a.25.25 0 0 1-.25.25H3.25a.25.25 0 0 1-.25-.25V5.25A.25.25 0 0 1 3.25 5ZM7 16.75V7H5v14h10v-2H7.25a.25.25 0 0 1-.25-.25v-2Z" clipR ...
+      </svg>
+    </InfoTooltip>
+    <Toast.Root open={open} onOpenChange={setOpen}
```
we already have the `Banner` component so either we should just use it for now or refactor this existing `Banner` component to use `Toast` from radix-ui (instead of MUI `Snackbar`) if we want to move forward with it

aaronc commented on web-components/src/utils/styles/cn.ts:
```diff
@@ -0,0 +1,6 @@
+import { ClassValue, clsx } from 'clsx';
+import { twMerge } from 'tailwind-merge';
+
+export function cn(...inputs: ClassValue[]) {
```
@flagrede I'm not actually sure what this is for and when I should use it. Can you add some documentation?

flagrede replied:
Added some documentation here: [52ce3b7](https://github.com/regen-network/regen-web/pull/2167/commits/52ce3b76d43379c1d034d064e7c229dab80ad0b1)

It allows to override class names. For example, if you have this:
`const Button => (className) => <button className={cn('text-primary', className)}`
then calling
`<Button className="text-secondary" />`
will override text-primary by text-secondary

aaronc commented on web-components/src/components/icons/CopyIcon.tsx:
```diff
@@ -0,0 +1,7 @@
+export const CopyIcon = (props: React.SVGProps<SVGSVGElement>) =>
+  <svg width="24" height="24" fill="none" {...props}>
+    <path fill="#000" fillRule="evenodd" d="M19 8v9H9.25a.25.25 0 0 1-.25-.25V3h5v4.25c0 .414.336.75.75.75H19Zm-4-4.586L18.287 7H15V3.414ZM3.25 5H7V1.25A.25.25 0 0 1 7.25 1h8.14c.07 0 .137.03.184.081l5.36 5.847a.25.25 0 0 1 .066.17V18.75a.25.25 0 0 1-.25.25H17v3.75a.25.25 0 0 1-.25.25H3.25a.25.25 0 0 1-.25-.25V5.25A.25.25 0 0 1 3.25 5ZM7 16.75V7H5v14h10v-2H7.25a.25.25 0 0 1-.25-.25v-2Z" clipRule= ...
+  </svg>
```
It's not necessary to use the MUI `SvgIcon` - this works fine with tailwind

aaronc commented:
I think I've addressed all the remaining issues, just waiting for updated tailwind color names from @erikalogie.

blushi commented on web-marketplace/src/components/organisms/UserAccountSettings/UserAccountSettings.ConnectField.tsx:
```diff
@@ -0,0 +1,57 @@
+ * display a single social or wallet connection.
+ */
+export const ConnectField = ({
+                        providerName,
+                        connect,
+                        disconnect,
+                        address,
+                      }: {
```
some formatting issues still here
before I remember we ran linter check on commit, but it's been disabled because it was taking too much time, maybe we should add it to CI in a follow-up, wdyt?

blushi reviewed (approved):
tACK, nice work! pending some nits and fixing the TS error with the `PaletteOptions`: https://github.com/regen-network/regen-web/actions/runs/6499025237/job/17651518237?pr=2167

aaronc commented:
> tACK, nice work! pending some nits and fixing the TS error with the `PaletteOptions`: https://github.com/regen-network/regen-web/actions/runs/6499025237/job/17651518237?pr=2167

How did this TS error get introduced? I didn't edit this file and it's not failing locally

aaronc commented:
I have updated the tailwind color naming here to reflect @erikalogie's latest naming in https://www.figma.com/file/Md1Cl2B9eJi4NMzQgpbdMF/Dark-Mode?type=design&node-id=112-9885&mode=design&t=XSnmQmEBgHarcE9x-0

aaronc commented:
The TS error is now gone 🤷‍♂️. I'm going to go ahead and merge this

blushi commented:
> The TS error is now gone 🤷‍♂️. I'm going to go ahead and merge this

It looks like this came from this commit https://github.com/regen-network/regen-web/pull/2167/commits/fb09903f55452dc14d93045bd19c7328b5908566, which updated web-components/lib to web-components/src, and caused our [MUI `PaletteOptions` override](https://github.com/regen-network/regen-web/blob/dev/web-components/src/types/overrides-mui.d.ts) to not be taken into account.

flagrede commented:
> > The TS error is now gone 🤷‍♂️. I'm going to go ahead and merge this
> 
> It looks like this came from this commit [fb09903](https://github.com/regen-network/regen-web/commit/fb09903f55452dc14d93045bd19c7328b5908566), which updated web-components/lib to web-components/src, and caused our [MUI `PaletteOptions` override](https://github.com/regen-network/regen-web/blob/dev/web-components/src/types/overrides-mui.d.ts) to not be taken into account. But there's some way to go around that if we want to just use web-components/src/: https://stackoverflow.com/questions/68482092/material-ui-module-augmentation-across-multiple-packages-in-a-monorepo

Fixed that here: https://github.com/regen-network/regen-web/pull/2181/commits/23c467303abfc0be4fcd78903b20844077d29aeb
